src/irmin/proof.ml

```diff
@@ -71,125 +89,364 @@ struct
     let t elt_t = Type.map [%typ: (H.t * elt) list] of_list to_list
   end
 
-  (* Keep track of read effects happening during a computation using
-     sets. This does not keep track of the ordering of the reads. *)
-  type read_set = { nodes : N.t Hashes.t; contents : C.t Hashes.t }
+  type mode = Produce | Serialise | Deserialise | Consume [@@deriving irmin]
+  type kind = Structural | Streamed
+
+  module Structural = struct
```

Comment:
Not very fond of `Structural` here; `Tree`? Or `Set`?

Reply:
I don't understand `Set` 🤔 . I like `Tree` but I feel this would be confusing regarding `tree.ml`.

`Inductive`? Or `Unordered`?

Reply:
`Set` as we are tracking read effects as a set vs. a stream bellow.